lightningd/peer_htlcs.c

```diff
@@ -867,21 +841,30 @@ htlc_accepted_hook_callback(struct htlc_accepted_hook_payload *request,
 				 channel_resolve_reply, gr);
 		} else
 			handle_localpay(hin, hin->cltv_expiry, &hin->payment_hash,
-					*rs->amt_to_forward,
-					*rs->outgoing_cltv,
-					*rs->total_msat,
-					rs->payment_secret);
+					request->onion->amt_to_forward,
+					request->onion->outgoing_cltv,
+					*request->onion->total_msat,
+					request->onion->payment_secret);
 		break;
 	case htlc_accepted_fail:
 		log_debug(channel->log,
 			  "Failing incoming HTLC as instructed by plugin hook");
-		if ((failure_code & UPDATE) && rs->nextcase == ONION_END) {
-			log_broken(channel->log,
-				   "htlc_acccepted hook: Can't return failure %u on last hop!",
-				   failure_code);
-			failure_code = WIRE_TEMPORARY_NODE_FAILURE;
+		if (failure_code & UPDATE) {
+			if (rs->nextcase == ONION_END) {
+				log_broken(channel->log,
+					   "htlc_acccepted hook: Can't return failure %u on last hop!",
+					   failure_code);
+				failure_code = WIRE_TEMPORARY_NODE_FAILURE;
+			} else if (!request->onion) {
+				log_broken(channel->log,
+					   "htlc_acccepted hook: Can't return failure %u on undecodable onion!",
+					   failure_code);
+				failure_code = WIRE_TEMPORARY_NODE_FAILURE;
```

Comment:
How come we can't fail if we can't decode?

Reply:
We can't send an UPDATE, since that requires us to know next hop...